File: examples_rust/examples/example_basic.rs
use libzupt_rust;

fn main() {
    println!("{}", "=".repeat(60));
    println!("libzupt - Basic Encryption/Decryption Example (Rust)");
    println!("{}", "=".repeat(60));
    println!();

    println!("1. Generating key pair...");
    let keypair = libzupt_rust::generate_keypair().expect("Failed to generate key pair");
    println!("   Public key size: {} bytes", keypair.public_key.len());
    println!("   Secret key size: {} bytes", keypair.secret_key.len());
    println!();

    let message = b"Hello, Post-Quantum World! This is a secret message.";
    println!("2. Encrypting message: {}", String::from_utf8_lossy(message));
    let (ciphertext, enc_header) = libzupt_rust::encrypt(&keypair.public_key, message)
        .expect("Failed to encrypt");
    println!("   Ciphertext size: {} bytes", ciphertext.len());
    println!("   Header size: {} bytes", enc_header.len());
    println!();

    println!("3. Decrypting...");
    let decrypted = libzupt_rust::decrypt(&keypair.secret_key, &ciphertext, &enc_header)
        .expect("Failed to decrypt");
    println!("   Decrypted: {}", String::from_utf8_lossy(&decrypted));
    println!();

    assert_eq!(decrypted, message, "Decryption failed!");
    println!("4. Verification: SUCCESS - Decrypted message matches original");
    println!();

    println!("5. Testing with wrong key...");
    let keypair2 = libzupt_rust::generate_keypair().expect("Failed to generate key pair");
    match libzupt_rust::decrypt(&keypair2.secret_key, &ciphertext, &enc_header) {
        Ok(_) => println!("   ERROR: Should have failed!"),
        Err(e) => println!("   Correctly rejected with error: {}", e),
    }
    println!();

    println!("{}", "=".repeat(60));
    println!("All examples passed!");
    println!("{}", "=".repeat(60));
}

File: examples_go/example_basic.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"strings"
)

func runExampleBasic() error {
	sep := strings.Repeat("=", 60)
	fmt.Println(sep)
	fmt.Println("libzupt - Basic Encryption/Decryption Example")
	fmt.Println(sep)
	fmt.Println()

	fmt.Println("1. Generating key pair...")
	keygen := NewKeyGenerator()
	keypair, err := keygen.GenerateKeyPair()
	if err != nil {
		return err
	}
	fmt.Printf("   Public key size: %d bytes\n", len(keypair.PublicKey))
	fmt.Printf("   Secret key size: %d bytes\n", len(keypair.SecretKey))
	fmt.Println()

	encryptor, err := NewEncryptor(keypair.PublicKey)
	if err != nil {
		return err
	}
	decryptor, err := NewDecryptor(keypair.SecretKey)
	if err != nil {
		return err
	}

	message := []byte("Hello, Post-Quantum World! This is a secret message.")
	fmt.Printf("2. Encrypting message: %s\n", message)
	ciphertext, encHeader, err := encryptor.Encrypt(message)
	if err != nil {
		return err
	}
	fmt.Printf("   Ciphertext size: %d bytes\n", len(ciphertext))
	fmt.Printf("   Header size: %d bytes\n", len(encHeader))
	fmt.Println()

	fmt.Println("3. Decrypting...")
	decrypted, err := decryptor.Decrypt(ciphertext, encHeader)
	if err != nil {
		return err
	}
	fmt.Printf("   Decrypted: %s\n", decrypted)
	fmt.Println()

	if !bytes.Equal(decrypted, message) {
		return errors.New("Decryption failed!")
	}
	fmt.Println("4. Verification: SUCCESS - Decrypted message matches original")
	fmt.Println()

	fmt.Println("5. Testing with wrong key...")
	keygen2 := NewKeyGenerator()
	keypair2, err := keygen2.GenerateKeyPair()
	if err != nil {
		return err
	}
	decryptorWrong, err := NewDecryptor(keypair2.SecretKey)
	if err != nil {
		return err
	}

	if _, err := decryptorWrong.Decrypt(ciphertext, encHeader); err != nil {
		fmt.Printf("   Correctly rejected with error: %v\n", err)
	} else {
		fmt.Println("   ERROR: Should have failed!")
	}
	fmt.Println()

	fmt.Println(sep)
	fmt.Println("All examples passed!")
	fmt.Println(sep)
	return nil
}
